Support for building cythonized distribution archive

andy-maier opened the pull request:
This adds support so that users can build a cythonized version of the wheel distribution archive for pywbem, by using **either one** of these commands:
`$ make buildc`
or
`$ ./setup.py bdist_wheel --cythonized`

Note that `./setup.py bdist_wheel --help` displays help text that covers the `--cythonized` option.

The cythonized version can be successfully built on macOs and Ubuntu on Python 3.5 and higher, but not on Windows.
The failing tests should only be the occasional ConnectionError failures when sending indications.

See the commit message for details.

rchateauneu commented:
Great job.

Out of curiosity, I tried on Python 2.7 and Windows 7 . Maybe this is not pywbem preferred target, indeed, but it might be of interest.

It compiles with extremely few warnings, but there was a link error. As far as I remember, I had a similar problem in the past and this problem vanished when I excluded __init__.py files from compilation. This might not be a perfect solution but at least, it works. Log file and one-line patch to setup.py are attached.

[setup.patch.txt](https://github.com/pywbem/pywbem/files/6128507/setup.patch.txt)
[pywbem_build.log](https://github.com/pywbem/pywbem/files/6128373/pywbem_build.log)

At the end, I could not find anything in pywbem\build\bdist.win-amd64 ; Maybe this is intended as this platform is not supported but it seems perfectly doable.

FYI, the setup.py options "build_ext --inplace" , sometimes implemented, allows to compile and link Python files into pyd, yielding an immediate speedup and allowing to adjust Cython directives. I use it here: https://github.com/rchateauneu/survol/blob/master/setup.py

More tests to come.

rchateauneu commented:
This command:
```
python3 setup.py bdist_wheel --cythonized
```
... builds successfully on:
```
NAME="Ubuntu"
VERSION="20.04.2 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.2 LTS"
VERSION_ID="20.04"
```

... running on Windows Subsystem for Linux (WSL).

andy-maier commented:
@rchateauneu Remi, I have updated the PR to add documentation, and to have the cythinized version built as part of the scheduled tests and when releasing pywbem. A test or resource measurements have not been added at this point. Unless you insist on tests, I would go forward with this PR.

andy-maier commented:
A few more updates to the PR were needed to address basic issues with the installation of the cythonized archive.

At this point, the three environments that are capable of running the test all fail consistently with the same error that is raised in the ply module:

        if pinfo.error:
>           raise YaccError('Unable to build parser')
E           ply.yacc.YaccError: Unable to build parser

/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/ply/yacc.py:3282: YaccError

andy-maier commented:
I found the problem: The 'module' argument of 'yacc.yacc()' was not used, and the algorithm to determine the module automatically seems to skip cythonized call levels. Fixed that in an update to the PR, and added a second commit that improves the error handling of this situation for both yacc and lex.

andy-maier commented:
The last changes brought it to the point where only the occasional ConnectionError when sending indications shows up.
I think we are ready for review and merge now.

KSchopmeyer commented on docs/intro.rst:
```diff
@@ -238,6 +238,42 @@ can be installed.
+target system, this speeds up execution in compute-intensive areas of pywbem
+such as when parsing the CIM-XML responses of a WBEM server. The downside is
+that a cythonized wheel archive is significantly larger than a universal wheel
+archive, and the memory consumption is also larger.
```
propose change to "both the wheel archive  and memory consumption are significantly  larger that a universal wheel archive."

andy-maier replied:
Well, I did not measure the memory consumption, and while I think it must be larger based on the archive size, I would not necessarily call this out to be significant without having measured it. So I would like to have two separate comparison statements. NOT CHANGED.

andy-maier replied:
In the meeting today, Karl agreed to go without resolving this comment.

KSchopmeyer reviewed (commented):
One minor comment but that is a lot of code and must have been a lot of work.  I reviewed but did not test.

andy-maier commented:
I would like to not make your proposed change. See the comment for reasons. Let me know whether you're ok with that.
